### ISSUE TITLE
License of project?

### ISSUE THREAD

**Issue by SteveMacenski** [dheera/rosshow#16]
I'm compiling a set of tasks for migrating ROS2 and I think this is something good enough to think about moving into a core CLI repo after being converted to ROS2. 

To start that work, I'd need to know what license this is available under

**Comment by dheera** [comment]
Hi @SteveMacenski --

- I've licensed it under 3-clause BSD, and the LICENSE.txt file is under the rosshow/ directory.

- I realize it's a bit of an awkward repo structure -- since there is a lot in common between the ROS and ROS2 versions -- namely the entire src/librrosshow can be common to both, and only the "rosshow" executable needs to change. I have a preliminary ROS2 version not yet committed to the repo but I'm also trying to experiment with the possibility of a single package that is installable and functional in both ROS/ROS2, although not 100% certain if that's a good idea.

- I'd love for this to be part of the core CLI repo; let me know if the licensing needs to change for that to happen, and also if you have suggestions on ROS/ROS2 cross-compatibility while keeping librosshow common to both.

Thanks!

**Comment by SteveMacenski** [comment]
So I actually say the license, but then also in your package XML you have MIT [listed](https://github.com/dheera/rosshow/blob/master/rosshow/package.xml#L7), so that would need to be rectified. BSD lets me just go at it without needing to think twice, the MIT license I would need to contact some folks to sort out.

I've heard of some people trying cross installing but I don't know the state of it, I know it wasn't easy. I dont know what you're using internally but if its ROS2, I'd just recommend dropping ROS1 support and focusing on where things are going. I've ported all my repos to Melodic but I'm only accepting new PRs for the ROS2 branch unless I can trivially cherry pick over.

No promises but this seems useful and right up the alley for ros2cli.

**Comment by dheera** [comment]
@SteveMacenski Thanks for pointing that out! I just fixed package.xml which was in error and now reads BSD.

**Comment by SteveMacenski** [comment]
Fair enough, lets circle back on this when I get to this, it'll probably take a couple weeks as right now I'm scrambling to finish some other work to make time in October to write a roscon talk

**Comment by clalancette** [comment]
For what it is worth, I did a super-quick port of this to ROS 2: https://github.com/clalancette/rosshow/tree/ros2

What I ended up doing there is splitting the code up into `librosshow` and `rosshow2`.  Most of the functionality lives in `librosshow`, while `rosshow2` is more-or-less just the node wrapper around it.  I didn't do the port of `rosshow{1}`, but it should be straightforward.

One way we can go about supporting both ROS 1 and ROS 2 fairly easily is to make 3 separate repositories; one that contains `librosshow`, one that contains `rosshow`, and one that contains `rosshow2`.  For ROS 1 (like Melodic), you'd release `librosshow` and `rosshow`, and for ROS 2 you'd release `librosshow` and `rosshow2`.  Another way to do it would be to have different branches on this repository, which is what we are generally doing on other repositories.

Anyway, I probably will not have time to move the ROS 2 port forward myself, but I thought I'd point out what I did in case @SteveMacenski wants to start from that.

**Comment by SteveMacenski** [comment]
Awesome! Thanks for the starting point. 

On the release part: I had thought you could even keep them in the same repo if you add a COLCON_IGNORE file to the ros1 version and CATKIN_IGNORE to the ros2 and that way it’ll only build the right one for each- but I  recall reading somewhere the build farm doesn’t respect those, but I can’t find a citation for that and may not be true. 

With that said, I don’t think that’s probably the best way to go about it.

**Comment by dheera** [comment]
Hi @SteveMacenski just an update here -- I'm working on making rosshow dual ROS1/ROS2 using the same repo. It turns out everything can be exactly the same except for package.xml and CMakeLists.txt, so there are just 2 configure scripts to put them in place (see dev branch for latest work).

I do want to include these into the ROS build farms for both ROS1 and ROS2; is there any way to have the build farms run a configure script? If not, I think that might be fine as since the build farms seem to want an entirely separate repo anyway I can just clone/run/commit for the build farm purposes.

I'm also working on a web-browser based visualizer: https://github.com/dheera/rosboard/ that I will announce in discourse soon, as soon as I have at least LaserScan/Odometry/OccupancyGrid working.

**Comment by SteveMacenski** [comment]
I don't think so unless you can get OR to set up a custom build configuration for you.

**Comment by clalancette** [comment]
No, there is no way to run a configure script on the buildfarm.  That said, there are conditional depends in package.xml, and CMake is basically turing-complete, so you might be able to auto-detect the ROS version you are on.

**Comment by dheera** [comment]
> No, there is no way to run a configure script on the buildfarm. That said, there are conditional depends in package.xml, and CMake is basically turing-complete, so you might be able to auto-detect the ROS version you are on.

@clalancette 
Awesome thanks, I'll play with that and see if I can get it working. I'm also looking for a good reference on how to create ROS2 packages with CMake instead of relying on the `ament_python` magic. The ROS2 docs seem to not describe how to do this.

**Comment by clalancette** [comment]
> Awesome thanks, I'll play with that and see if I can get it working with conditionals in `package.xml`. I'm also looking for a good reference on how to create ROS2 Python-only packages with CMake instead of relying on the `ament_python` magic. The ROS2 docs seem to not describe how to do this. I got a package to build to `install/` but unless I use `ament_python`, `ros2 run` seems to not be able to actually find the node.

In general, we recommend to use `ament_python` for pure Python packages; it is the most supported thing.  What particular problems caused you to move away from `ament_python`?

**Comment by dheera** [comment]
@clalancette The main concerns I had, and maybe I'm wrong! -- feel free to point me to the right resources but I found the official documentation pretty sparse:

A. Installing packages to /opt/ros instead of using colcon for deployed production robots (usually I don't like to have anything in /home or in "workspaces" like /some_ws for deployments)

B. Packages that contain both Python and C++ nodes, for which ament_python doesn't take care of the C++, and CMake and ament_python may fight over who gets to do the Python part.

C. Building dual ROS1/ROS2 Python packages. ROS1 requires a CMakeLists.txt, and it seems ROS2 gets confused by both a CMakeLists.txt and package.xml containing ament_python. I was considering just doing only CMakeLists.txt with CMake conditionals. I might be doing something wrong though.
